File: src/main/java/se/gewalli/commands/AddProductToOrderCommand.java
package se.gewalli.commands;

import com.fasterxml.jackson.annotation.JsonProperty;
import se.gewalli.data.EntityNotFound;
import se.gewalli.data.Repository;
import se.gewalli.entities.Order;
import se.gewalli.entities.Product;

import java.util.ArrayList;

public record AddProductToOrderCommand(@JsonProperty("id") int id,
                                       @JsonProperty("version") int version,
                                       @JsonProperty("orderId") int orderId,
                                       @JsonProperty("productId") int productId) implements Command {
    @Override
    public CommandType getType() {
        return CommandType.AddProductToOrderCommand;
    }

    @Override
    public void run(Repository repository) throws EntityNotFound {
        var order = repository.getOrder(orderId);
        var productList = new ArrayList<Product>(order.products());
        productList.add(repository.getProduct(productId));
        repository.save(new Order(order.id(), order.customer(), order.orderDate(), productList, order.version() + 1));
    }
}

File: src/main/java/se/gewalli/commands/AddProductToOrderCommand.kt
package se.gewalli.commands

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import se.gewalli.data.EntityNotFound
import se.gewalli.data.Repository
import se.gewalli.entities.Order

class AddProductToOrderCommand @JsonCreator constructor(
    @JsonProperty("id") id: Int,
    @JsonProperty("version") version: Int,
    @param:JsonProperty("orderId") val orderId: Int,
    @param:JsonProperty("productId") val productId: Int
) : Command(id, version) {
    override val type: CommandType get()= CommandType.AddProductToOrderCommand

    @Throws(EntityNotFound::class)
    override fun run(repository: Repository) {
        val order = repository.getOrder(orderId)
        val productList = ArrayList(order.products)
        productList.add(repository.getProduct(productId))
        repository.save(Order(order.id, order.customer, order.orderDate, productList, order.version + 1))
    }
}